### PR TITLE
return ls as internal API

### LINKED ISSUE #478: Re-implement `mlem.api.ls`
We removed `mlem.api.ls` in #395 
Yet, having it (even with just a recursive search) would be good.
Few related issues we had in the past:
- https://github.com/iterative/mlem/issues/187
- https://github.com/iterative/mlem/issues/99
- https://github.com/iterative/mlem/issues/98
- https://github.com/iterative/mlem/issues/94

### PR DESCRIPTION
**Pull request by mike0sv**
close #478

### CODE REVIEW

**Comment by aguschin** [comment]
Wow, do you have any idea why CI didn't run? 🤔

**Comment by mike0sv** [comment]
No idea

File: mlem/core/metadata.py [hunk 1]
```diff
@@ -214,3 +221,28 @@ def find_meta_location(location: Location) -> Location:
...
+        if isinstance(meta, MlemLink):
+            type_ = meta.link_cls
+        else:
+            parent = meta.__parent__
+            if (
+                parent is not None
+                and parent != MlemObject
+                and issubclass(parent, MlemObject)
+            ):
+                type_ = parent
```
[the remark this answers is not in the document]
Comment (mike0sv) [reply mlem/core/metadata.py:246]: Dont recall any other places with such logic. 
So we can have 
1) a link - we put it into whatever it links to
2) a non-polymorphic object, which goes to it's type bucket
3) a polymorphic object, which goes to it's parent type bucket

[the remark this answers is not in the document]
Comment (omesser) [reply mlem/core/metadata.py:246]: I understand, just feels like messing with the classes' internals. So this type resolution is only specific to the `ls` implementation today?

[the remark this answers is not in the document]
Comment (aguschin) [reply mlem/core/metadata.py:246]: @mike0sv? :)

[the remark this answers is not in the document]
Comment (aguschin) [reply mlem/core/metadata.py:246]: > So this type resolution is only specific to the ls implementation today?

Yes.

**Review by aguschin — APPROVED** [review]

**Pull request merged**